File: sdk/java-sdk/src/main/java/kalix/javasdk/StatusCode.java
/*
 * Copyright 2021 Lightbend Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kalix.javasdk;


/**
 *  Interface used to represent a status code, typically used when replying with an error.
 *
 */

public interface StatusCode {
  /**
   * The supported HTTP error codes that can be used when replying from the Kalix user function.
   * Note: this list is not an exhaustive list of all the possible HTTP codes but only the ones available for use currently.
   *
   */
  enum HttpErrorCode implements StatusCode {
    BAD_REQUEST(400),
    UNAUTHORIZED(401),
    FORBIDDEN(403),
    NOT_FOUND(404),
    CONFLICT(409),
    TOO_MANY_REQUESTS(429),
    INTERNAL_SERVER_ERROR(500),
    SERVICE_UNAVAILABLE(503),
    GATEWAY_TIMEOUT(504);

    public final int httpCode;
    HttpErrorCode(int httpCode) {
      this.httpCode = httpCode;
    }
  }

}

File: sdk/scala-sdk/src/main/scala/kalix/scalasdk/StatusCode.scala
/*
 * Copyright 2021 Lightbend Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kalix.scalasdk

import kalix.javasdk.StatusCode.{ ErrorCode => JErrorCode }
import kalix.javasdk.{ StatusCode => JStatusCode }

sealed trait ErrorCode extends JStatusCode {}
object BadRequest extends ErrorCode
object Unauthorized extends ErrorCode
object Forbidden extends ErrorCode
object NotFound extends ErrorCode
object Conflict extends ErrorCode
object TooManyRequests extends ErrorCode
object InternalServerError extends ErrorCode
object ServiceUnavailable extends ErrorCode
object GatewayTimeout extends ErrorCode

private[scalasdk] object StatusCodeConverters {
  def toJava(statusCode: ErrorCode): JErrorCode = {
    statusCode match {
      case BadRequest          => JErrorCode.BAD_REQUEST
      case Unauthorized        => JErrorCode.UNAUTHORIZED
      case Forbidden           => JErrorCode.FORBIDDEN
      case NotFound            => JErrorCode.NOT_FOUND
      case Conflict            => JErrorCode.CONFLICT
      case TooManyRequests     => JErrorCode.TOO_MANY_REQUESTS
      case InternalServerError => JErrorCode.INTERNAL_SERVER_ERROR
      case ServiceUnavailable  => JErrorCode.SERVICE_UNAVAILABLE
      case GatewayTimeout      => JErrorCode.GATEWAY_TIMEOUT
    }
  }
}
